File: leetcode-scala/src/dfs/ConvertSortedArrayToBinarySearchTree.scala
package dfs

import tools.TreeNode

object ConvertSortedArrayToBinarySearchTree {

    /**
     * 1152 ms    65.00%
     * 58.3 MB    100.00%
     * @param nums
     * @return
     */
    def sortedArrayToBST(nums: Array[Int]): TreeNode = {
        recursiveHelper(nums, 0, nums.length - 1)
    }

    def recursiveHelper(nums: Array[Int], start: Int, end: Int): TreeNode = {
        if (start > end) return null

        val mid = (start + end) >> 1
        val root = new TreeNode(nums(mid))
        root.left = recursiveHelper(nums, start, mid - 1)
        root.right = recursiveHelper(nums, mid + 1, end)

        root
    }

    def main(args: Array[String]): Unit = {
        val treeNode = sortedArrayToBST(Array.range(0, 10))
        print(treeNode.toString)
    }
}

File: leetcode-java/src/classify/binary_search_tree/easy/ConvertSortedArrayToBinarySearchTree.java
package classify.binary_search_tree.easy;

import tools.TreeNode;

public class ConvertSortedArrayToBinarySearchTree {
    
    /**
     * 0ms      100.00%
     * 41.6MB   11.2%
     */
    public TreeNode sortedArrayToBST(int[] nums) {
        return construct(nums, 0, nums.length - 1);
    }

    private TreeNode construct(int[] nums, int left, int right) {
        if (left > right) {
            return null;
        }

        if (left == right) {
            return new TreeNode(nums[left]);
        }

        int mid = left + (right - left) / 2;
        TreeNode node = new TreeNode(nums[mid]);
        node.left = construct(nums, left, mid - 1);
        node.right = construct(nums, mid + 1, right);

        return node;
    }
}